### PR TITLE
[WIP] Simplify architecture

### PR DESCRIPTION
**Pull request by shredlocker**
# Why?

The architecture is a bit complicated. I just wanted to see how much work it was to potentially refactor it. This is a suggestion for a straight forward architecture. Not planning on merging this one!

# What?

- Removed view controller factory and navigator classes.
- Removed dependancy container replaced with dependency injection.
- Refactored FilterViewController as a superclass instead of container view controller.

I'm just throwing this one out here. Should work as it used to, but I might have missed something at this point. Just wanted to see how much work it was to remove these things and if this is something we should change. I think this setup could make the KAFs a bit easier to implement as well, if we use the delegate pattern we discussed at some point.

# Show me

Should be the same as before

### CODE REVIEW

File: Sources/MultiLevelListSelectionFilter/MultiLevelListSelectionFilterViewController.swift [hunk 1]
```diff
@@ -124,7 +121,10 @@ public final class MultiLevelListSelectionFilterViewController: UIViewController
     }
 
     private func didSelectDrillDownItem(_ listItem: MultiLevelListSelectionFilterInfoType, at indexPath: IndexPath) {
-        filterSelectionDelegate?.filterContainerViewController(filterContainerViewController: self, navigateTo: listItem)
```
[the remark this answers is not in the document]
Comment (shredlocker) [reply Sources/MultiLevelListSelectionFilter/MultiLevelListSelectionFilterViewController.swift]: These two lines are equal with the current implementation in my opinion, it is not really delegation when you specifically tell the receiver what to do with the list item. But do I agree that it should be a delegate call and that some object should handle it and navigate accordingly :)

[the remark this answers is not in the document]
Comment (vadymmarkov) [reply Sources/MultiLevelListSelectionFilter/MultiLevelListSelectionFilterViewController.swift]: Yeah, it shouldn't be `navigate`, but something like `didSelect`.

**Comment by osanoj** [comment]
Part of this was already done in https://github.com/finn-no/charcoal-ios/pull/113, right?

Generally I agree with what @vadymmarkov wrote:
- NavigatorFactory seems unnecessary, we will not be switching between different navigation objects. So make a separate PR to remove that whenever you want 😄 
- `FilterDependencyContainer` and `ViewControllerFactory` can make sense to have since it is managing some dependencies when creating view controllers. But maybe having a protocol `ViewControllerFactory` that is implemented by `FilterDependencyContainer` is making it overly complicated. It looks like you planned to replace them with a new `FilterNavigationController` that holds dependencies (but doesn't seem to use it?). Perhaps something in that direction is better, with a "navigation" object that holds the dependencies and creates view controllers. Not sure if there is any advantage in mixing it with subclassing UINavigationController though 🤔

**Comment by shredlocker** [comment]
The subclass of UINavigationController is just a wrapper around the root filter. I wanted to make the
root filter controller a navigation controller, but wasn't sure how to do that the correct way :P 

If I was about to use this library in my own app I would expect it to be something like this:
```swift
let filter = Filter(data: someData)
let filterViewController = FilterViewController(filter: filter)
filterViewController.delegate = self
```
where Filter is a data structure that is defined in Charcoal and I would like to be able to use the filter like this:
```swift
let urlString = endpoint + filter.urlEncoded
// or
let data = filter.jsonEncoded
```

**Pull request closed without merging**